protocols/observed/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+//! to report the externally visible address (as listener).
+
+extern crate bytes;
+extern crate futures;
+extern crate libp2p_core;
+extern crate tokio_codec;
+extern crate tokio_io;
+extern crate unsigned_varint;
+
+use bytes::Bytes;
+use futures::{future, prelude::*};
+use libp2p_core::{ConnectionUpgrade, Endpoint, Multiaddr};
+use std::{io, iter};
+use tokio_codec::{FramedRead, FramedWrite};
+use tokio_io::{AsyncRead, AsyncWrite};
+use unsigned_varint::codec::UviBytes;
+
+/// The output, this connection upgrade produces.
+pub enum Output<C> {
+    /// As `Dialer`, we get our own externally observed address.
+    Address(Multiaddr),
+    /// As `Listener`, we return a sender which allows reporting the observed
+    /// address the client.
+    Sender(Sender<C>)
+}
+
+/// `Sender` allows reporting back the observed address to the remote endpoint.
+pub struct Sender<C> {
+    io: FramedWrite<C, UviBytes>
+}
+
+impl<C: AsyncWrite> Sender<C> {
+    /// Send address `a` to remote as the observed address.
+    pub fn send_address(self, a: &Multiaddr) -> impl Future<Item=C, Error=io::Error> {
+        self.io.send(Bytes::from(a.to_bytes())).map(|io| io.into_inner())
+    }
+}
+
+/// The connection upgrade type to retrieve or report externally visible addresses.
+pub struct Observed {}
+
+impl Observed {
+    pub fn new() -> Self {
+        Observed {}
+    }
+}
+
+impl<C> ConnectionUpgrade<C> for Observed
+where
+    C: AsyncRead + AsyncWrite + Send + 'static
+{
+    type NamesIter = iter::Once<(Bytes, Self::UpgradeIdentifier)>;
+    type UpgradeIdentifier = ();
+    type Output = Output<C>;
+    type Future = Box<dyn Future<Item=Self::Output, Error=io::Error> + Send>;
+
+    fn protocol_names(&self) -> Self::NamesIter {
+        iter::once((Bytes::from("/paritytech/observed-address/0.1.0"), ()))
+    }
+
+    fn upgrade(self, conn: C, _: (), role: Endpoint) -> Self::Future {
+        match role {
+            Endpoint::Dialer => {
+                let io = FramedRead::new(conn, UviBytes::default());
+                let future = io.into_future()
+                    .map_err(|(e, _): (io::Error, FramedRead<C, UviBytes>)| e)
+                    .and_then(move |(bytes, _)| {
+                        if let Some(b) = bytes {
+                            let ma = Multiaddr::from_bytes(b.to_vec())
+                                .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
+                            Ok(Output::Address(ma))
+                        } else {
+                            Err(io::ErrorKind::InvalidData.into())
+                        }
+                    });
+                Box::new(future)
+            }
+            Endpoint::Listener => {
+                let io = FramedWrite::new(conn, UviBytes::default());
+                Box::new(future::ok(Output::Sender(Sender { io })))
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    extern crate tokio_current_thread;
```

Comment:
Since there's already a PR to replace this with `tokio` (#577), I think you should do it here as well.